smac/optimizer/acquisition/maximizer.py

```diff
@@ -455,17 +455,25 @@ def _do_search(
                 if not active[i]:
                     continue
                 if obtain_n[i] == 0 or improved[i]:
-                    obtain_n[i] = 2
+                    obtain_n[i] = self.vectorization_min_obtain
                 else:
                     obtain_n[i] = obtain_n[i] * 2
                     obtain_n[i] = min(obtain_n[i], self.vectorization_max_obtain)
                 if new_neighborhood[i]:
                     if not improved[i] and n_no_plateau_walk[i] < self.n_steps_plateau_walk:
-                        if len(neighbors_w_equal_acq[i]) != 0:
+                        if len(neighbors_w_equal_acq[i]) > 0:
                             candidates[i] = neighbors_w_equal_acq[i][0]
                             neighbors_w_equal_acq[i] = []
+                        else:
+                            # Local optima: All neighbours are worse
```

Comment:
I am not so sure about that. Assuming that we have lots of hyperparameters, as each time we only sample 8 neighbors, it is quite likely that all these 8 neighbors are all worse than the current one while the acquisition function values can be further reduced with the other neighbors

Reply:
I follow your line of reasoning, but I would argue that this concern should be addressed in the neighbourhood generator (by, for example, increasing the numbers of neighbours to sample) and not in the local search procedure. In fact, we primarily sample more than one neighbourhood value to increase the chance of getting an improvement, right? Or are there other motivations for doing this? 

Furthermore, before the local search was vectorized in (v0.11.0), the search did also stop when it reached a local optimum.

Reply:
So this basically disables the plateau walking, right?

One would assume that increasing the number of neighbors would do the same thing as disabling plateau walking. While I agree, the issue would be that looking at more neighbors at the same time, this would increase the overhead of SMAC because these extra points would need to go through the random forest and acquisition function.

Maybe @frank-hutter can provide further intuition as he proposed this functionality back then.

Reply:
Plateau walk still occurs as long as there are neighbours with the same acquisition value as the current candidate. Only when all neighbours have an acquisition value that is worse than the value of the current candidate, I propose to consider it as a local optimum and stop the search. 
In the previous implementation the neighbourhood of the same current candidate will be sampled (which also include categorical and ordinal parameters where the neighbourhood sampling is deterministic (right?)). This also results in an increased overhead of (guaranteed futile) computations per plateau walk. Furthermore, considering you actually are in a local optimum, is it worthwhile to do 10 plateau walks (assuming the default value and no previous plateau walks in the search trajectory) before stopping the search?

Reply:
Interesting points. There are two things I extract from this:
1. Plateau walk must be disabled for candidates that return all neighbors. I'm afraid that this needs to be done on the level of the config space?
2. We need to benchmark this to see whether it improves performance or not. @alexandertornede you mentioned that you are working on a benchmarking setup in some other context?